### PR TITLE
feat: user can control or not the search value in Transfer Component

### PR DESCRIPTION
**Pull request by Leoguedesm1**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

 [#17240](https://github.com/ant-design/ant-design/issues/17240)

### 💡 Background and solution

Added a new prop in Transfer Component with target and source list input search values and the update functions.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   Added a new prop in Transfer Component with target and source list input search values and the update functions.        |
| 🇨🇳 Chinese |  在 Transfer Component 中添加了一个新道具，其中包含目标和源列表输入搜索值和更新功能         |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
```jsx
showSearch?: boolean | {
  searchValue?: string;
  onSearchValueChange?: (value: string) => void;
}
```

**Comment by Leoguedesm1** [comment]
> ```jsx
> showSearch?: boolean | {
>   searchValue?: string;
>   onSearchValueChange?: (value: string) => void;
> }
> ```

@afc163 i don't get it. How can I control both search values separately with one variable?

**Comment by afc163** [comment]
```jsx
showSearch?: boolean | [{
  searchValue?: string;
  onSearchValueChange?: (value: string) => void;
}, {
  searchValue?: string;
  onSearchValueChange?: (value: string) => void;
}]
```

File: components/transfer/index.en-US.md [hunk 1]
```diff
@@ -34,8 +34,7 @@ One or more elements can be selected from either column, one click on the proper
 | render | The function to generate the item shown on a column. Based on an record (element of the dataSource array), this function should return a React element which is generated from that record. Also, it can return a plain object with `value` and `label`, `label` is a React element and `value` is for title | (record) => ReactNode | - |  |
 | selectAllLabels | A set of customized labels for select all checkboxs on the header | (ReactNode \| (info: { selectedCount: number, totalCount: number }) => ReactNode)\[] | - |  |
 | selectedKeys | A set of keys of selected items | string\[] | \[] |  |
-| showSearch | If included, a search box is shown on each column | boolean | false |  |
-| search | If included, you can control the search input value | object \| ({targetValue: string, onChangeTargetValue: (searchValue: string => void), sourceValue: string, onChangeSourceValue: (searchValue: string => void)}) | - |  |
+| showSearch | If included, a search box is shown on each column | boolean \| \[ { searchValue: string, onSearchValueChange: (value: string) => void }, { searchValue: string, onSearchValueChange: (value: string) => void } \] |  | false |  |
```
Comment (afc163) [inline components/transfer/index.en-US.md:37]:
```suggestion
| showSearch | If included, a search box is shown on each column | boolean \| \[{ searchValue: string, onSearchValueChange: (value: string) => void }, { searchValue: string, onSearchValueChange: (value: string) => void }\] |  | false |  4.20.0: Array<{ searchValue, onSearchValueChange }> |
```

**Comment by afc163** [comment]
Could you help to add some test cases?

**Comment by Leoguedesm1** [comment]
> Could you help to add some test cases?

i will try, i never used jest, only cypress

**Comment by Leoguedesm1** [comment]
@afc163 I can't thinking in more situations, i added only a test case, tell me if this is enough or if you can tell me more situations to test

File: components/transfer/__tests__/search.test.js [hunk 2]
```diff
@@ -89,4 +89,16 @@ describe('Transfer.Search', () => {
       .simulate('change', { target: { value: ' ' } });
     expect(filterOption).toHaveBeenCalledTimes(dataSource.length);
   });
+
+  it('typing value should trigger showSearch.onSearchValueChange', () => {
+    const onSearchValueChange = jest.fn();
+    const wrapper = mount(
+      <Transfer dataSource={dataSource} showSearch={{ searchValue: '', onSearchValueChange }} />,
```
Comment (afc163) [inline components/transfer/__tests__/search.test.js:96]: Should it be?

```suggestion
      <Transfer dataSource={dataSource} showSearch={[{ searchValue: '', onSearchValueChange }]} />,
```

Comment (Leoguedesm1) [reply components/transfer/__tests__/search.test.js]: @afc163 it's done

**Comment by li-jia-nan** [comment]
Hey! Conflicting files, and it seems this is now really outdated. If this is something that still makes sense to the latest master version of antd, please send a new PR with this change :)

**Pull request closed without merging**